### PR TITLE
Typed Protocols Redesign

### LINKED ISSUE #3334: Rewrite of typed-protocols

### PR DESCRIPTION
**Pull request by coot**
Fixes input-output-hk/cardano-node#4221.

### CODE REVIEW

File: ouroboros-network/ouroboros-network.cabal [hunk 1]
```diff
@@ -222,11 +222,8 @@ library
 library protocol-tests
   hs-source-dirs:      protocol-tests
   default-language:    Haskell2010
-  exposed-modules:     Ouroboros.Network.Protocol.BlockFetch.Direct
-                       Ouroboros.Network.Protocol.BlockFetch.Examples
+  exposed-modules:     Ouroboros.Network.Protocol.BlockFetch.Examples
```
Comment (bolt12) [inline ouroboros-network/ouroboros-network.cabal:225]: can this commit ( commit c9eb7f1918fc9a6b27774e949272b3f83dd541c1 ) be squashed with all other `ouroboros-protocol-tests: updated against typed-protocols changes` commits?

File: docs/interface-CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,23 @@ team.  See [consensus
 CHANGELOG](../ouroboros-consensus/docs/interface-CHANGELOG.md) file for how
 this changelog is supposed to be used.
 
+## Circa 2022-05-23
+
+- redesign of `typed-protocols`.  Each pipelining client now tracks unsatisfied
+  transitions using a type level fifo. For example in the context of a simple
+  [`ping-poing` mini-protocol](https://input-output-hk.github.io/ouroboros-network/typed-protocols-examples/Network-TypedProtocol-PingPong-Type.html#t:PingPong)
```
Comment (bolt12) [inline docs/interface-CHANGELOG.md:12]:
```suggestion
  [`ping-pong` mini-protocol](https://input-output-hk.github.io/ouroboros-network/typed-protocols-examples/Network-TypedProtocol-PingPong-Type.html#t:PingPong)
```

File: docs/interface-CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,23 @@ team.  See [consensus
...
+  if a pipelining clients sends two `MsgPing`, the queue will contain two yet
+  unsatisfied transitions `StBusy -> StIdle`.  They can be taken out of the
+  queue in a fifo order using one of the `Collect`, `CollectSTM`, `CollectDone`
+  primitives which usually are packaged.  In particular all pipelining clients
+  will need to be updated.
+
+- `Data.Type.Nat` was moved from `typed-protocols` to `ouroboros-network` package.
+
+- Pipeliend decisions are still using `Nat` singletons; one needs
```
Comment (bolt12) [inline docs/interface-CHANGELOG.md:21]:
```suggestion
- Pipelined decisions are still using `Nat` singletons; one needs
```

File: ouroboros-consensus-test/ouroboros-consensus-test.cabal [hunk 4]
```diff
@@ -99,6 +99,7 @@ library
                      , quiet             >=0.2   && <0.3
                      , random
                      , serialise         >=0.2   && <0.3
+                     , singletons
```
Comment (bolt12) [inline ouroboros-consensus-test/ouroboros-consensus-test.cabal:102]: This commit (6f660ccc5e8944379897df866fec8f4810ff7c7b) as well (squashed)

**Review by bolt12 — COMMENTED** [review]
It's a bit hard to review such large diffs, but most of them are refactoring to make the code type check after the typed-protocols redesign. If it would make sense to squash most of the commits related to that, I think it would be easier to look at the patches and infer what's worth more attention.

File: network-mux/src/Network/Mux/Channel.hs [hunk 5]
```diff
@@ -168,6 +194,43 @@ socketAsChannel socket =
         then return Nothing
         else return (Just (LBS.fromStrict chunk))
 
+    tryRecv :: IO (Maybe (Maybe LBS.ByteString))
+#if defined(mingw32_HOST_OS)
+    -- TODO
+    tryRecv = return Nothing
+#else
```
[the remark this answers is not in the document]
Comment (coot) [reply network-mux/src/Network/Mux/Channel.hs:201]: This requires non-blocking IO on windows, I left it as a future improvement. I think I should transform the todo into an issue.

File: network-mux/src/Network/Mux/Types.hs [hunk 6]
```diff
@@ -228,15 +228,16 @@ newtype SDUSize = SDUSize { getSDUSize :: Word16 }
 -- 'MiniProtocolNum' and 'MiniProtocolDir'.
 --
 muxBearerAsChannel
-  :: forall m. Functor m
+  :: forall m. Applicative m
```
[the remark this answers is not in the document]
Comment (coot) [reply network-mux/src/Network/Mux/Types.hs:231]: we need `pure` or `return` in the dummy `tryRecv` implementation.

File: ouroboros-network/ouroboros-network.cabal (hunk 1 above)
Comment (coot) [reply ouroboros-network/ouroboros-network.cabal:225]: There's only one commit which updates `ouroboros-protocol-tests`.

File: ouroboros-consensus-test/ouroboros-consensus-test.cabal (hunk 4 above)
Comment (coot) [reply ouroboros-consensus-test/ouroboros-consensus-test.cabal:102]: I prefer smaller commits which updates a single component at a time, which allows to focus more on smaller chunks related to each other.

File: ouroboros-network/ouroboros-network.cabal (hunk 1 above)
Comment (coot) [reply ouroboros-network/ouroboros-network.cabal:225]: Oh I think I know what you mean, yes it makes sense to squash the two commits.

**Pull request closed without merging**